### PR TITLE
Pack the struct to save significant gas

### PR DESCRIPTION
**Pull request by omnus**
A full uint256 isn't needed for the dates and the use case (unless we have more use cases than atoms in the universe :) ).

Save six storage slots per delegation by packing these as uint96s next to each address (160 + 96 = 256).

Gas before this change is 541,555 (https://goerli.etherscan.io/tx/0xd44e28fe97d73008e8924cce53c74dcdb08b41d229d51a337c1fcad9f8d24ced)

After the change is 408,183 (https://goerli.etherscan.io/tx/0x2b470a99b18a61ce41440b86f80a53ea627e7294d3fc6210f95d890737700b50)

### CODE REVIEW

**Review by omnus — COMMENTED** [review]
Gah, solidity formatter in VS code makes it look like I changed everything, lol. I did not.

(Can recommend this formatter if you don't have one: solidity extension by Juan Blanco).

**Comment by estarriolvetch** [comment]
Is there any significant difference than #4?

**Comment by rookmate** [comment]
> Is there any significant difference than #4?

Looks mostly the same to me if the goal was to re-org the struct to save gas

File: src/DelegationManagement.sol [hunk 1]
```diff
@@ -1,183 +1,317 @@
...
  *  @date: 21-Dec-2022 @ 10:30
- *  @version: 4.23 
+ *  @version: 4.23
  *  @notes: This is a experimental contract for delegation registry
  *  @author: skynet2030 (skyn3t2030)
  *
  */
 
-pragma solidity >=0.6.0 <=0.8.0;
+pragma solidity 0.8.17;
```
Comment (rookmate) [inline src/DelegationManagement.sol]: Break this change into a different commit. You're "just" not packing the struct differently. You are also changing the compiler to the latest version. This should be a separate commit

File: src/DelegationManagement.sol [hunk 2]
```diff
@@ -1,183 +1,317 @@
...
+    mapping(bytes32 => delegationAddresses[]) public delegateToHashes;
+    mapping(bytes32 => delegationAddresses[]) public delegateFromHashes;
 
     // Events declaration
 
-    event registerDelegation(address indexed from, address indexed collectionAddress, address indexed delegationAddress, uint256 useCase);
-    event revokeDelegation(address indexed from, address indexed collectionAddress, address indexed delegationAddress, uint256 useCase);
-    event updateDelegation(address indexed from, address indexed collectionAddress, address olddelegationAddress, address indexed newdelegationAddress, uint256 useCase);
-    
+    event registerDelegation(
```
Comment (rookmate) [inline src/DelegationManagement.sol]: Beautifying the event declarations should also be a different commit

File: src/DelegationManagement.sol [hunk 3]
```diff
@@ -1,183 +1,317 @@
...
+     *
      */
-    function registerDelegationAddress(address _collectionAddress, address _delegationAddress, uint256 _expiryDate, uint256 _useCase) public {
-        require((_useCase >0 && _useCase < useCaseCounter) || (_useCase == 99));
+    function registerDelegationAddress(
+        address _collectionAddress,
+        address _delegationAddress,
+        uint96 _expiryDate,
+        uint96 _useCase
+    ) public {
```
Comment (rookmate) [inline src/DelegationManagement.sol]: Remove beautified code from this commit

**Review by rookmate — CHANGES_REQUESTED** [review]
Beautifying the code should be a different commit than changing the packing struct.

Your PR name and single commit are misleading to the actual changes your code suggest.

Please split your commits so that others may understand all the separate changes you've submitted.

**Comment by omnus** [comment]
> Is there any significant difference than #4?

Nothing hugely significant, exactly the same aim. My change is very slightly more gas efficient (about 56 gas per registration, which yes is tiny, but also non-0).

* My version: https://goerli.etherscan.io/tx/0xb7a01637955568aa03e14f8fee75c623d2be873041108c6c725add54dde0ae22 408,159
* Yours: https://goerli.etherscan.io/tx/0x5eeb3589cacabd2f4d43154604ef2f247d07ffd9da7eea19e7071115cd67b1a3 408,215

I believe this is down to casting to a larger int (uint96), than the smaller uint64 taking less effort. TBH I am not entirely sure, but have always found that using a larger unit and packing to the 256 limit of the slot is slightly more efficient (thank goodness for uint96!).

(But yeah, either solution is a big improvement! :) )

File: src/DelegationManagement.sol (hunk 3 above)
Comment (omnus) [reply src/DelegationManagement.sol]: Fair point. Done!

File: src/DelegationManagement.sol (hunk 2 above)
Comment (omnus) [reply src/DelegationManagement.sol]: I agree

File: src/DelegationManagement.sol (hunk 1 above)
Comment (omnus) [reply src/DelegationManagement.sol]: Sure. Done!

**Review by rookmate — APPROVED** [review]
